sockeye/inference.py

```diff
@@ -511,35 +535,151 @@ def get_max_output_length(input_length: int):
+
+
+def make_input_from_plain_string(sentence_id: int, string: str) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a plain string.
+
+    :param sentence_id: An integer id.
+    :param string: An input string.
+    :return: A TranslatorInput.
+    """
+    return TranslatorInput(sentence_id, tokens=list(data_io.get_tokens(string)), factors=None)
+
+
+def make_input_from_json_string(sentence_id: int, json_string: str) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a JSON object, serialized as a string.
+
+    :param sentence_id: An integer id.
+    :param json_string: A JSON object serialized as a string.
+    :return: A TranslatorInput.
+    """
+    jobj = json.loads(json_string, encoding=C.JSON_ENCODING)
+    tokens = jobj.get(C.JSON_TEXT_KEY)
+    if tokens is None:
+        raise ValueError("Missing 'text' field in JSON string: %s", json_string)
+
+    tokens = list(data_io.get_tokens(tokens))
+    factors = jobj.get(C.JSON_FACTORS_KEY)
+    if isinstance(factors, list):
+        factors = [list(data_io.get_tokens(factor)) for factor in factors]
+    else:
+        factors = None
+
+    return TranslatorInput(sentence_id=sentence_id, tokens=tokens, factors=factors)
+
+
+def make_input_from_factored_string(sentence_id: int,
+                                    factored_string: str,
+                                    translator: 'Translator',
+                                    delimiter: str = C.DEFAULT_FACTOR_DELIMITER) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a string with factor annotations on a token level, separated by delimiter.
+    If translator does not require any source factors, the string is parsed as a plain token string.
+
+    :param sentence_id: An integer id.
+    :param factored_string: An input string with additional factors per token, separated by delimiter.
+    :param translator: A translator object.
+    :param delimiter: A factor delimiter. Default: '|'.
+    :return: A TranslatorInput.
+    """
+    utils.check_condition(bool(delimiter) and not delimiter.isspace(),
+                          "Factor delimiter can not be whitespace or empty.")
+    model_num_factors = translator.num_source_factors
+    num_additional_factors = model_num_factors - 1
+
+    if num_additional_factors == 0:
+        return make_input_from_plain_string(sentence_id=sentence_id, string=factored_string)
+
+    tokens = []  # type: Tokens
+    factors = [[] for _ in range(num_additional_factors)]  # type: List[Tokens]
+    for token_id, token in enumerate(data_io.get_tokens(factored_string)):
+        token, *token_factors = token.rsplit(delimiter, maxsplit=num_additional_factors)
+        utils.check_condition(bool(token), "Empty token at sentence {}, position {}".format(sentence_id, token_id))
+
+        token_factors = list(filter(None, token_factors))
+        utils.check_condition(len(token_factors) == num_additional_factors and all(token_factors),
```

Comment:
Wouldn't all(token_factors) always be true once all Nones are filtered out?

Reply:
but they can be empty as well which evaluates to False as well for all()

Reply:
I was getting confused, assuming that an empty token would result in None instead of empty string. The all() makes perfect sense then. But why the filter then? When can a None appear in token_factors?

Reply:
If you remove the filter, this test fails:
https://github.com/awslabs/sockeye/pull/275/files#diff-06d3a83a115dacea5dce8b1ab9a024adR197

This is because `rsplit()` creates an empty element in the resulting list:
```
In [5]: "test|".rsplit('|')
Out[5]: ['test', '']

In [6]: "test|".rsplit('|', maxsplit=1)
Out[6]: ['test', '']

In [7]: "test|".rsplit('|', maxsplit=2)
Out[7]: ['test', '']
```

Reply:
Ok, I was getting confused again. So filter(None, ...) does filter empty lists. Then the comment below gets active again!